bn254/src/helpers.rs

```diff
@@ -242,88 +240,264 @@ pub(crate) fn halve_bn254(mut input: [u64; 4]) -> [u64; 4] {
+    let a_signed = if s_f == -1 { a_sub } else { a };
+    let (b_sub, _) = wrapping_sub(BN254_PRIME, b);
+    let b_signed = if s_g == -1 { b_sub } else { b };
+
+    let product = linear_comb_unsigned(a_signed, b_signed, abs_f, abs_g);
+    interleaved_monty_reduction(product[0], product[1..].try_into().unwrap())
+}
+
+/// An adjustment factor equal to `2^{1030} mod P`
+pub(crate) const BN254_2_POW_1030: [u64; 4] = [
+    0x1f7ca21e7fcb111b,
+    0x61a09399fcfe8a6c,
+    0x1438cc5aab55aedb,
+    0x020c9ba0aeb6b6c7,
+];
+
+/// Invert a value in the BN254 field using a GCD based inversion algorithm.
+///
+/// The following approach to a GCD based inversion algorithm is taken from the paper:
+/// Optimized Binary GCD for Modular Inversion by Thomas Pornin.
+/// It can be found here: https://eprint.iacr.org/2020/972.pdf
+//
+/// Explicitly, we implement Algorithm 2, roughly following the C implementation linked in that paper (in section 3).
+/// The algorithm is a variant of the Binary Extended Euclidean Algorithm which, allows for most of the iterations to
+/// be performed using only u64's even in cases where the inputs are much larger.
+//
+/// This implementation is also impervious to side-channel attacks as an added bonus. In principal we could make
+/// the average case a little faster if we didn't care about this property but the worst case would be unchanged and
+/// potentially even slightly worse.
+pub(crate) fn gcd_inversion(val: [u64; 4]) -> [u64; 4] {
+    // When u = 1 and v 0, this outputs a^{-1} mod P.
+    // More generally, when u = K, this output a^{-1}K mod P. Hence if we can choose the right
+    // initial value for u, we can save some ourselves some work.
+    // We start with aR so this will output (aR)^{-1} whereas we want a^{-1}R. Hence we need to multiply
+    // by R^2 = 2^{512}
+    // We do 31 * 15 + 41 = 506 iterations. Each iteration injects a power of 2 so we need to divide by 2^{506}.
+    // Each pair of calls to linear_comb_monty_red, involves dividing by `2^64`. Hence to correct for all these
+    // we need to multiply by `2^{16 * 64}`.
+    // Overall we want `u` to equal `2^{16 * 64 + 6} = 2^{1030} mod P`
+    let (mut a, mut u, mut b, mut v) = (val, BN254_2_POW_1030, BN254_PRIME, [0, 0, 0, 0]);
+
+    // We need 506 iterations as in each iteration, all we can guarantee is that len(a) + len(b) will decrease by 1.
+    // Initially, `len(a) + len(b) <= 2 * 254 = 508` so we need 506 iterations to get to the point that the sum of the
+    // lengths is 2. Once it is 2, we know both `a` and `b` must be `1` or `0` as neither can be `0` without the other being
+    // `1` due to the fact that the GCD is `1`.
+    // We split the iterations into 15 initial rounds of size 31 and a final round of size 41.
+    const ROUND_SIZE: usize = 31; // If you want to change round size, you will also need to modify the constant in rm_middle.
+    const FINAL_ROUND_SIZE: usize = 41;
+    for _ in 0..15 {
```

Comment:
Nit: Should give number of rounds a name, and can check the constants for free.
```suggestion
    const N_ROUNDS: usize = 15;
    const ROUND_SIZE: usize = 31; // If you want to change round size, you will also need to modify the constant in rm_middle.
    const FINAL_ROUND_SIZE: usize = 41;
    assert_eq!(N_ROUNDS * ROUND_SIZE + FINAL_ROUND_SIZE, 506);
    for _ in 0..N_ROUNDS {
```